### ISSUE TITLE
Possibly unhandled error should be thrown

### ISSUE THREAD

**Issue by ysmood** [jakearchibald/es6-promise#70]
Node v0.10, non-harmony. Such as:

```javascript
var Promise = require('bluebird');
new Promise(function () { throw('err') });
// Unhandled error will be thrown.
```

```javascript
var Promise = require('es6-promise').Promise;
new Promise(function () { throw('err') });
// Nothing will happen.
```

**Comment by stefanpenner** [comment]
Upstream RSVP supports this, as this isn't part of the spec it is up to @jakearchibald if he wants it or not.

Let me know, and i'll add it.

**Comment by ysmood** [comment]
This should be part of the shim for old browser. Or it could be really error-prone in practical usage.
For the sake of the Promise spec, we could add an option to turn it on or off.

**Comment by stefanpenner** [comment]
I strangely suspect @jakearchibald  will be fine with this.

**Comment by benjamingr** [comment]
I've written the following proposal about this here https://gist.github.com/benjamingr/0237932cee84712951a2 I'd love your feedback. Here's the related RSVP issue: https://github.com/tildeio/rsvp.js/issues/355

**Comment by mziwisky** [comment]
+1, throw me them errors, please -- spent an hour tracking down what turned out to be a missing `require` call because the ReferenceError that was thrown during the success callback got swallowed up and hidden from my prying eyes.  makes me very sad.

at the least, es6-promise could expose a sort of catchall error-handling API as RSVP seems to do here: https://github.com/tildeio/rsvp.js#error-handling

**Comment by jakearchibald** [comment]
Chrome currently logs unhandled errors. Firefox does it on GC.

Happy to do the same as Chrome here. It should be a log rather than a throw, we'd be spec-breaking if it appeared in `window.onerror`.

**Comment by ysmood** [comment]
@jakearchibald node.js still don't throw anything.

**Comment by jakearchibald** [comment]
Yep

**Comment by benjdlambert** [comment]
IMO its a totally different type of error, its not a synchronous error, its not like you can wrap a try-catch block round it.

It should be something that should be pushed into the .catch handler, as the promise spec states, maybe logged similar to how bluebird doing it.

**Comment by benjamingr** [comment]
@ysmood io is merging a patch for the events (my proposal) in the next version. I think you should implement this

**Comment by ollym** [comment]
Is this going to be done? A `console.error(err)` is all we need.
Really difficult to debug without it.

**Comment by ysmood** [comment]
@ollym I wrote another Promsie lib, it's smaller, faster, and more error friendly. For more details goto: https://github.com/ysmood/yaku

**Comment by stefanpenner** [comment]
@ysmood good job! Im curious if you have any enumeration/denodify benchmark examples to go with your existing benchmarks.

**Comment by ysmood** [comment]
@stefanpenner I'm not sure what the `enumeration/denodify benchmark` means. Is it means non-promise `callback version` benchmark?

**Comment by stefanpenner** [comment]
benchmarks involving `Promise.all`, or benchmarks involving assimilation  of node callback style API.

**Comment by spikebrehm** [comment]
Is there something we project consumers can do to help move this along?  This is a bit painful.

**Comment by ysmood** [comment]
@stefanpenner You can replace `promise-es6` with [yaku](https://github.com/ysmood/yaku). Their api are the same, and `yaku` fixed the problem.

**Comment by ysmood** [comment]
@spikebrehm You can replace promise-es6 with yaku. Their api are the same, and yaku fixed the problem.

**Comment by OliverJAsh** [comment]
when does this according to @domenic’s spec: https://github.com/cujojs/when/blob/master/docs/debug-api.md#global-rejection-events

Events:

```
- 'unhandledRejection': fired when an unhandled rejection is detected
- 'rejectionHandled': fired when rejection previously reported via an 'unhandledRejection' event becomes handled
```

The default event handler should log.

**Comment by sirpengi** [comment]
I'm using v2.2.0 (installed via bower from github.com/components) and I got around this issue with the following shim:

```
Promise.prototype._onerror = function(err) {
    console.assert(false, err);
}
```

I don't see why this won't also work on the latest

**Comment by ysmood** [comment]
I think I should close this issue, because I solved the problem perfectly with [Yaku](https://github.com/ysmood/yaku), and I can't find any reason to keep using es6-promise any more.

**Comment by usmonster** [comment]
@ysmood in the interest of the greater community that does continue to use and rely on this polyfill, perhaps reopen the issue until it is actually resolved?

@jakearchibald, any comment on the intent to implement this?

Thanks!

**Comment by ysmood** [comment]
@usmonster This issue has lasted half year now. There no API difference between most ES6 promise shim libs. You simply change one line of the `package.json` of your project, everything will go well. Why do you waste time to wait?

**Comment by frosas** [comment]
@ysmood it's not only about waiting for @jakearchibald to address it, but to give visibility of the issue to anyone considering to use this library.

**Comment by ysmood** [comment]
@frosas You are right. If I am the author of ES6-promise, I will write a list to compare different promise libs on the README, and let people aware of the key disadvantages of each lib, include es6-promise itself. We should give people the chance to choose, not hidden the truth.

**Comment by Niondir** [comment]
+1 
Chrome also throws the error when not catched. Please do so in es6-promise.

**Comment by domenic** [comment]
Chrome does not throw an error.

**Comment by benjamingr** [comment]
Right, I think what @Niondir wants is to add the behavior you specified for Chrome 49 with `window.addEventListener("unhandledrejection"` - I definitely would like to see es6-promise do this.

**Comment by mmoss** [comment]
What I'm currently seeing.. It definitely appears that Chrome **does** throw an error...

**Chrome Native Promise without Rejection Handler:**
![nativepromiserethrownwithoutrejectionhandler](https://cloud.githubusercontent.com/assets/648160/11932701/568d1f8a-a7c5-11e5-9ddd-8ab47063b528.png)

**Transcompiled using Babel without a Rejection Handler in Chrome:**
_No indication that something's gone awry_

![promiseshimwithoutrejectionhandlerswallowserror](https://cloud.githubusercontent.com/assets/648160/11932700/54886e24-a7c5-11e5-962b-93745baa81d0.png)

**Comment by Niondir** [comment]
I also do not know why this polyfill takes effect in browsers where promises are implemented already. An alternative solution coud be to just not define the Polyfill where Promise already exists.

**Comment by benjamingr** [comment]
A lot of browsers implement the spec incorrectly and do not allow, for example, to subclass promises.

This behavior can be easily amended by implementing the unhandled rejection spec (both the browser and Node) and `console.error`ing on unhandled rejection like in Chrome.

**Comment by mmoss** [comment]
@benjamingr are you suggesting a simple monkey patch when using the es6-promise library, or a fix to the lib itself?

**Comment by ggarek** [comment]
The means to catch unhandled rejections is a part of [ECMAScript 2017 Draft spec (as HostPromiseRejectionTracker)](https://tc39.github.io/ecma262/#sec-host-promise-rejection-tracker). Looking forward to see it in this library!

**Comment by stefanpenner** [comment]
> The means to catch unhandled rejections is a part of ECMAScript 2017 Draft spec (as HostPromiseRejectionTracker). Looking forward to see it in this library!

Ya, still trying to find time ;)

**Comment by ljharb** [comment]
To clarify, the only thing in the ES spec is implementation-specific hooks, and restrictions around what implementations may do. An example implementation of this hook is in [whatwg](https://html.spec.whatwg.org/multipage/webappapis.html#unhandled-promise-rejections), but that's not part of ECMAScript itself.

**Comment by nilliams** [comment]
Hi, would love to use this polyfill, but been avoiding for this reason.

Now that Node has a [workaround](https://news.ycombinator.com/item?id=12737382) for this, any chance this polyfill can also buck the (IMO lacking) standard and do so too?

**Comment by stefanpenner** [comment]
@nilliams it will be part of 5.0, as described by the milestone assignment

**Comment by nilliams** [comment]
Okay that is great news thanks!

(And sorry I don't find those sorts of github 'events' that interject the issue thread particularly explicit, my eyes tend to skip over them).

**Comment by nilliams** [comment]
I want to be open about the fact I've recommended the isomorphic-fetch module moves away from this as suggested polyfill for this reason, until it's fixed, though it's fairly likely I'm just out of the loop on this: https://github.com/matthew-andrews/isomorphic-fetch/issues/111

**Comment by stefanpenner** [comment]
@nilliams I only have so many cycles, given infinite time this would have been done already ;)

**Comment by nilliams** [comment]
Yes I know and appreciate that, and apologies ... I really don't mean to pick on your project. I am just very aware that both the `fetch` project and the `isomorphic-fetch` are recommending using es6-promise as the default polyfill and I feel it really isn't safe to use yet.

I've just added an even worse example to the other thread, I think that bug is actually a separate issue in your tracker -- I think there's a workaround involving using setTimeout in `.catch()` but I don't have the notes at hand right now: https://github.com/matthew-andrews/isomorphic-fetch/issues/111#issuecomment-255443016

**Comment by stefanpenner** [comment]
> (There are other reasons, like - last I checked - the promise.prototype.finally polyfill also doesn't work with it).

please see: https://github.com/stefanpenner/es6-promise/issues/232#issuecomment-246454813 
implementation:  https://github.com/stefanpenner/es6-promise/pull/233

Just waiting patiently to pull the trigger.

**Comment by stefanpenner** [comment]
> I just really don't feel it is safe to use yet, and I feel like at the moment I'm running around telling people 'hey don't use that cool fetch thing just yet', and currently this is a reason why.

This isn't new information, I personally use a variant of this exact library which has this feature ;) I just haven't had spare cycles to port it over. Hopefully soon!

**Comment by stefanpenner** [comment]
Locking this, because no new information is possible. The constraint has been the same for some time, and that is my own time (or someones contribution to this project).

At some-point hopefully I have the spare cycles, until then please be mindful of this.
